src/Development/IDE/Plugin/CodeAction.hs

```diff
@@ -375,6 +380,132 @@ suggestSignature isQuickFix Diagnostic{_range=_range@Range{..},..}
+--   Possible fix: add (Eq b) to the context of the instance declaration
+-- • In the second argument of ‘(&&)’, namely ‘x' == y'’
+--   In the expression: x == y && x' == y'
+--   In an equation for ‘==’:
+--       (Pair x x') == (Pair y y') = x == y && x' == y'
+  | Just constraint <- findMissingConstraint _message
+  , Just [instanceLineStr, constraintFirstCharStr]
+    <- matchRegex _message "bound by the instance declaration at .+:([0-9]+):([0-9]+)"
+  = let existingConstraints = secondLine _message
+        newConstraints = normalizeConstraints existingConstraints constraint
+        instanceLine = readPositionNumber instanceLineStr
+        constraintFirstChar = readPositionNumber constraintFirstCharStr
+        startOfConstraint = Position instanceLine constraintFirstChar
+        endOfConstraint = Position instanceLine $
+          constraintFirstChar + T.length existingConstraints
+        range = Range startOfConstraint endOfConstraint
+        title = "Add `" <> constraint <> "` to the context of the instance declaration"
+     in [(title, [TextEdit range newConstraints])]
+  | otherwise = []
+    where
+      secondLine :: T.Text -> T.Text
+      secondLine = T.lines >>> flip (!!) 1 >>> T.strip >>> T.replace "from the context: " ""
+
+      readPositionNumber :: T.Text -> Int
+      readPositionNumber = T.unpack >>> read >>> pred
+
+findTypeSignatureName :: T.Text -> Maybe T.Text
+findTypeSignatureName t = matchRegex t "\\n\\s*(.+) :: " <&> head
+
+findTypeSignatureLine :: T.Text -> T.Text -> Int
+findTypeSignatureLine contents typeSignatureName =
+  T.splitOn (typeSignatureName <> " :: ") contents & head & T.lines & length
+
+-- | Suggests a constraint for a type signature for which a constraint is missing.
+suggestFunctionConstraint :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
+suggestFunctionConstraint contents Diagnostic{..}
+-- Suggests a constraint for a type signature with any number of existing constraints.
+-- • No instance for (Eq a) arising from a use of ‘==’
+--   Possible fix:
+--     add (Eq a) to the context of
+--       the type signature for:
+--         eq :: forall a. a -> a -> Bool
+-- • In the expression: x == y
+--   In an equation for ‘eq’: eq x y = x == y
+
+-- • Could not deduce (Eq b) arising from a use of ‘==’
+--   from the context: Eq a
+--     bound by the type signature for:
+--                eq :: forall a b. Eq a => Pair a b -> Pair a b -> Bool
+--     at Main.hs:5:1-42
+--   Possible fix:
+--     add (Eq b) to the context of
+--       the type signature for:
+--         eq :: forall a b. Eq a => Pair a b -> Pair a b -> Bool
+-- • In the second argument of ‘(&&)’, namely ‘y == y'’
+--   In the expression: x == x' && y == y'
+--   In an equation for ‘eq’:
+--       eq (Pair x y) (Pair x' y') = x == x' && y == y'
+  | Just c <- contents
+  , True <- _message =~ ("the type signature for:" :: String)
+  , Just constraint <- findMissingConstraint _message
```

Comment:
`findMissingConstraint _message` can get evaluated up to 3 times, which seems a bit wasteful. It would be good to reorganize things so that it only needs to be called once.

Reply:
Good point! I'll try to address this.

Reply:
I've created an entrypoint for missing constraint suggestion in order to perform `findMissingConstraint _message` only once and then routing to the relevant implementation.